src/context.ts

```diff
@@ -0,0 +1,59 @@
+import React, {
+  createContext as reactCreateContext,
+  useContext,
+  useRef,
+} from 'react'
+import create, { UseStore } from 'zustand'
+import { EqualityChecker, State, StateCreator, StateSelector } from './vanilla'
+
+function createContext<TState extends State>(initialState?: TState) {
+  const ZustandContext = reactCreateContext<UseStore<TState> | undefined>(
+    undefined
+  )
+
+  const Provider = ({
+    createState,
+    children,
+  }: {
+    createState: StateCreator<TState>
+    children: React.ReactNode
+  }) => {
+    const storeRef = useRef<UseStore<TState>>()
+
+    if (!storeRef.current) {
+      storeRef.current = create(createState)
```

Comment:
It's only important this is initialized once at first, so we could technically accept the store (=useStore).
```js
      storeRef.current = useStore // from Provider's props
```
Is this more handy? or confusing?

Reply:
@dai-shi Ah.. createState prop on Provider alone is not going to work for another reason.

Have a look at my last [next.js solution](https://codesandbox.io/s/nextjs-with-zustand-forked-sr6z4?file=/lib/store.js:1266-1450) useHydrate() hook. (It is a working example, with all the bugs fixed I mentioned at #182).

See the logic to merge new states into existing store on client-side? It is unavoidable, and has to be there for next.js to work properly. Our current Provider interface is not doing this, nor does it expose the store.

So I am in favor of passing store. Property name can be either `store` or maybe `initialStore`?

Reply:
I'm not sure if I fully understand the issue/solution in nextjs, but passing store is fine. I prefer `initialStore` too, or even `initialUseStore` because it's `useStore` from create, not `store` from vanilla.

Reply:
I'll call it `initialStore` for simplicity, as.. we also have a `useStore` as return value of `createContext()`. It could get confusing with two useStore's in application code (yes we know internally they are two different things).

Reply:
Okay. Please check other comments too: https://github.com/pmndrs/zustand/pull/375#pullrequestreview-649818401